### PR TITLE
feat(codec): Introduce `Decoder/Encoder` traits

### PR DESCRIPTION
**Pull request by alce**
In preparation for experimenting with a different decoding strategy, this PR swaps tokio's `Decoder` and `Encoder` traits with our own.

### CODE REVIEW

File: tonic/src/codec/encode.rs [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
 use bytes::{BufMut, Bytes, BytesMut};
 use futures_core::{Stream, TryStream};
 use futures_util::{ready, StreamExt, TryStreamExt};
 use http::HeaderMap;
 use http_body::Body;
 use pin_project::pin_project;
-use std::pin::Pin;
-use std::task::{Context, Poll};
-use tokio_util::codec::Encoder;
+
```
Comment (LucioFranco) [inline tonic/src/codec/encode.rs]: just for style can we keep the deps all grouped together?

File: tonic/src/codec/decode.rs [hunk 2]
```diff
@@ -18,16 +18,22 @@ const BUFFER_SIZE: usize = 8 * 1024;
 ///
 /// This will wrap some inner [`Body`] and [`Decoder`] and provide an interface
 /// to fetch the message stream and trailing metadata
-pub struct Streaming<T> {
-    decoder: Box<dyn Decoder<Item = T, Error = Status> + Send + Sync + 'static>,
+pub struct Streaming<T, D>
+where
+    D: Decoder<Item = T, Error = Status> + Send + Sync + 'static,
```
Comment (LucioFranco) [inline tonic/src/codec/decode.rs]: you don't need to add this bound here

**Comment by LucioFranco** [comment]
@alce friendly ping, anything I can help here?

**Comment by alce** [comment]
Hi, I’ll resume working on this early next week.

File: tonic/src/codec/buffer.rs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                    front.advance(rem);
+                    cnt -= rem;
+                }
+            }
+            self.bufs.pop_front();
+        }
+    }
+}
+
+impl<'a> Buf for DecodeBuf<'a> {
```
Comment (LucioFranco) [inline tonic/src/codec/buffer.rs]: I think you can elide this lifetime and just do `DecodeBuf<'_>`

File: tonic/src/codec/mod.rs [hunk 4]
```diff
@@ -37,3 +41,29 @@ pub trait Codec: Default {
...
+    /// The type of items consumed by the `Encoder`
+    type Item;
+
+    /// The type of encoding errors.
+    ///
+    /// The type of unrecoverable frame encoding errors.
+    type Error: From<io::Error>;
+
+    /// Encodes a frame into the buffer provided.
+    fn encode(&mut self, item: Self::Item, dst: &mut BytesMut) -> Result<(), Self::Error>;
```
Comment (LucioFranco) [inline tonic/src/codec/mod.rs]: I think for consitency we should do the same here for encode https://docs.rs/tower-grpc/0.1.1/src/tower_grpc/generic/codec.rs.html#135-137 and add an EncodeBuf which wraps bytesmut

File: tonic/src/codec/prost.rs [hunk 5]
```diff
@@ -70,19 +70,14 @@ impl<U: Message + Default> Decoder for ProstDecoder<U> {
     type Item = U;
     type Error = Status;
 
-    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
-        let mut cursor = std::io::Cursor::new(&buf[..]);
+    // TODO: improve once we upgrade prost
```
Comment (LucioFranco) [inline tonic/src/codec/prost.rs]: https://github.com/hyperium/tonic/pull/226

**Comment by LucioFranco** [comment]
So I brought this branch inline with latest master and ran benchmarks:

```
 name              old ns/iter        new ns/iter        diff ns/iter  diff %  speedup
 chunk_size_100    967 (1039 MB/s)    1,169 (859 MB/s)            202  20.89%   x 0.83
 chunk_size_1005   368 (2730 MB/s)    394 (2550 MB/s)              26   7.07%   x 0.93
 chunk_size_500    481 (2089 MB/s)    532 (1889 MB/s)              51  10.60%   x 0.90
 message_count_1   364 (1387 MB/s)    393 (1284 MB/s)              29   7.97%   x 0.93
 message_count_10  1,551 (3255 MB/s)  1,860 (2715 MB/s)           309  19.92%   x 0.83
 message_count_20  2,882 (3504 MB/s)  3,436 (2939 MB/s)           554  19.22%   x 0.84
 message_size_10k  2,140 (9350 MB/s)  2,129 (9398 MB/s)           -11  -0.51%   x 1.01
 message_size_1k   512 (3925 MB/s)    543 (3701 MB/s)              31   6.05%   x 0.94
 message_size_5k   1,140 (8780 MB/s)  1,206 (8300 MB/s)            66   5.79%   x 0.95
```

Old is #6673f9 and new is this branch with a merge of that commit.

It looks like it may not be worth it or we should spend some more time on this.

**Comment by magnet** [comment]
I haven't dug into this much, but while updating to tonic 0.1 this has bitten us as it seems this is preventing me from doing zero-copy handling of Bytes like we used to do by hiding the Bytes/BytesMut type. I previously used tokio's BytesCodec and passed buffers (containing a flatbuffers payload) to tonic in a zero-copy manner.  Has this ability been lost?

**Comment by magnet** [comment]
Here's the raw byte codec I was using on the tonic alpha.

**Comment by davedbase** [comment]
@magnet did you find a solution to this?

**Comment by magnet** [comment]
I haven't, but I have been busy with other stuff since.

**Comment by davedbase** [comment]
Any chance you are going to continue with it @magnet? This is a pretty awesome addition. I'm assuming that the new trait design totally prohibits you though. Would Tonic need changes to make this possible?

**Comment by magnet** [comment]
Not in the short term, no, maybe I'll pick this up sometimes next year.

**Pull request merged**